### PR TITLE
New PM sensor Panasonic SN-GCJA5

### PR DESCRIPTION
**Pull request by gcormier**
# What does this implement/fix?

Support for Panasonic SN-GCJA5 https://api.pim.na.industrial.panasonic.com/file_stream/main/fileversion/244938

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#3011

## Test Environment

- [x] ESP32
- [ ] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040

## Example entry for `config.yaml`:
```
esp32:
  board: esp32-c3-devkitm-1
  framework:
    type: arduino

uart:
  rx_pin: GPIO10
  parity: EVEN
  baud_rate: 9600

esphome:
  name: pannypm

sensor:
  - platform: gcja5
    pm_1_0:
      name: "Particulate Matter <1.0µm Concentration"
    pm_10_0:
      name: "Particulate Matter <10.0µm Concentration"
    pm_2_5:
      name: "Particulate Matter <2.5µm Concentration"
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

**Comment by functionpointer** [comment]
Not a maintainer, but i think the `core` label is bad for a PR like this.
It means you are modifying files that are part of the esphome core, which shouldn't be needed for a new sensor.

Looks like you changed something the `clang-format` script. Is that really needed?

**Comment by gcormier** [comment]
As far as the label, I didn't apply it, bot did.

I removed my clang-format changes.

File: tests/test1.yaml [hunk 1]
```diff
@@ -340,6 +344,14 @@ mcp23s17:
...
 
 sensor:
+  - platform: gcja5
+    pm_1_0:
+      name: "Particulate Matter <1.0µm Concentration"
+    pm_10_0:
+      name: "Particulate Matter <10.0µm Concentration"
+    pm_2_5:
+      name: "Particulate Matter <2.5µm Concentration"
+    uart_id: gcja5_uart
```
Comment (jesserockz) [inline tests/test1.yaml]: Please add all sensors here so we know if they get broken in the future.

File: esphome/components/gcja5/gcja5.cpp [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+    ESP_LOGI(TAG, "Overall Status : %i", (status >> 6) & 0x03);
+    ESP_LOGI(TAG, "PD Status      : %i", (status >> 4) & 0x03);
+    ESP_LOGI(TAG, "LD Status      : %i", (status >> 2) & 0x03);
+    ESP_LOGI(TAG, "Fan Status     : %i", (status >> 0) & 0x03);
+  }
+}
+
+void GCJA5Component::dump_config() { ; }
+
+void GCJA5Component::update() {
```
Comment (jesserockz) [inline esphome/components/gcja5/gcja5.cpp]: You should update the sensors when the data comes in via the loop, not in update as this is not a polling sensor but just receives data when it is new and current.

**Review by jesserockz — CHANGES_REQUESTED** [review]

File: tests/test1.yaml (hunk 1 above)
Comment (gcormier) [reply tests/test1.yaml]: Completed

File: esphome/components/gcja5/gcja5.cpp (hunk 2 above)
Comment (gcormier) [reply esphome/components/gcja5/gcja5.cpp]: Is this what you had in mind?

https://github.com/esphome/esphome/pull/4988/files/e15550ae4e7f7bcff09f48fd5a0f6c2d2675f23a..0daf400e5baa4817c0cb3166a8e1b6463950e420#diff-eea3104323ebfe4453602fae493747639ce9353e41ea843605531ac92c284515R37-R65

File: esphome/components/gcja5/gcja5.cpp [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+  uint32_t pm2_5 = get_32_bit_uint(5);
+  uint32_t pm10_0 = get_32_bit_uint(9);
+  uint16_t pmc0_5 = get_16_bit_uint(13);
+  uint16_t pmc1_0 = get_16_bit_uint(15);
+  uint16_t pmc2_5 = get_16_bit_uint(17);
+  uint16_t pmc5_0 = get_16_bit_uint(21);
+  uint16_t pmc7_5 = get_16_bit_uint(23);
+  uint16_t pmc10_0 = get_16_bit_uint(25);
+  */
+
```
Comment (jesserockz) [inline esphome/components/gcja5/gcja5.cpp]:
```suggestion
```

**Review by jesserockz — DISMISSED** [review]
Yeah looks good now apart from this block of commented code

**Comment by ssieb** [comment]
@functionpointer it's definitely possible for a new sensor to be marked as a core change if there are new constants added to the const.py file.

File: esphome/components/gcja5/gcja5.cpp [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+              this->rx_message_[i]);
+  }
+
+  if (this->rx_message_[0] != 0x02 || this->rx_message_[31] != 0x03 || !this->calculate_checksum_()) {
+    ESP_LOGVV(TAG, "Discarding bad packet - failed checks.");
+    return;
+  } else
+    ESP_LOGVV(TAG, "Good packet found.");
+
+  have_good_data_ = true;
```
Comment (jesserockz) [inline esphome/components/gcja5/gcja5.cpp]: Please make sure to prefix the member variables with `this->`
```suggestion
  this->have_good_data_ = true;
```

**Review by jesserockz — CHANGES_REQUESTED** [review]

File: esphome/components/gcja5/gcja5.cpp [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+  uint8_t val;
+  while (this->available() != 0) {
+    this->read_byte(&val);
+    this->rx_message_.push_back(val);
+
+    // check if rx_message_ has 32 bytes of data
+    if (this->rx_message_.size() == 32) {
+      this->parse_data_();
+
+      if (have_good_data_) {
```
Comment (jesserockz) [inline esphome/components/gcja5/gcja5.cpp]:
```suggestion
      if (this->have_good_data_) {
```

File: esphome/components/gcja5/gcja5.cpp [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+
+  if (this->rx_message_[0] != 0x02 || this->rx_message_[31] != 0x03 || !this->calculate_checksum_()) {
+    ESP_LOGVV(TAG, "Discarding bad packet - failed checks.");
+    return;
+  } else
+    ESP_LOGVV(TAG, "Good packet found.");
+
+  this->have_good_data_ = true;
+  uint8_t status = this->rx_message_[29];
+  if (!first_status_log_) {
```
Comment (jesserockz) [inline esphome/components/gcja5/gcja5.cpp]:
```suggestion
  if (!this->first_status_log_) {
```

**Review by jesserockz — CHANGES_REQUESTED** [review]

**Review by jesserockz — APPROVED** [review]

**Pull request merged**